Fix flaky CI checks on Ubuntu

The current pytest runs on Ubuntu often fail during PR checks. It seems that sometimes the optimization steps take too long and time out:

```
=========================== short test summary info ============================
FAILED cirq-core/cirq/experiments/z_phase_calibration_test.py::test_calibrate_z_phases[angles6-error6-characterization_flags6] - RuntimeError: Optimal parameters not found: The maximum number of function evaluations is exceeded.
= 1 failed, 19825 passed, 121 skipped, 89 xfailed, 7 xpassed, 843 warnings in 334.79s (0:05:34) =
Error: Process completed with exit code 1.
```

I'm seeing this kind of thing happen on more than 50% of my PRs.

--- a/cirq-core/cirq/experiments/xeb_fitting.py
+++ b/cirq-core/cirq/experiments/xeb_fitting.py
@@ -655,7 +655,7 @@ def _fit_exponential_decay(
             cycle_depths,
             fidelities,
             p0=(a_0, layer_fid_0),
-            bounds=((0, 0), (1, 1))
+            bounds=((0, 0), (1, 1)),
         )
         try:
             return curve_fit()

--- a/cirq-core/cirq/experiments/xeb_fitting.py
+++ b/cirq-core/cirq/experiments/xeb_fitting.py
@@ -14,6 +14,7 @@
 """Estimation of fidelity associated with experimental circuit executions."""
 import dataclasses
 from abc import abstractmethod, ABC
+from functools import partial
 from typing import Dict, Iterable, List, Optional, Sequence, Tuple, TYPE_CHECKING
 
 import numpy as np
@@ -647,14 +648,24 @@ def _fit_exponential_decay(
     layer_fid_0 = np.clip(np.exp(slope), 0, 1)
     a_0 = np.clip(np.exp(intercept), 0, 1)
 
-    try:
-        (a, layer_fid), pcov = optimize.curve_fit(
+    def do_curve_fit():
+        curve_fit = partial(
+            optimize.curve_fit,
             exponential_decay,
             cycle_depths,
             fidelities,
             p0=(a_0, layer_fid_0),
-            bounds=((0, 0), (1, 1)),
+            bounds=((0, 0), (1, 1))
         )
+        try:
+            return curve_fit()
+        except RuntimeError:  # pragma: no cover
+            # Curve_fit didn't find a solution. Try once more w/ higher maxfev.
+            # Default (in SciPy v.1.14) is 100*(1+len(p0)) = 300 for our p0.
+            return curve_fit(maxfev=1000)
+
+    try:
+        (a, layer_fid), pcov = do_curve_fit()
     except ValueError:  # pragma: no cover
         return 0, 0, np.inf, np.inf
 

--- a/cirq-core/cirq/experiments/xeb_fitting.py
+++ b/cirq-core/cirq/experiments/xeb_fitting.py
@@ -14,7 +14,6 @@
 """Estimation of fidelity associated with experimental circuit executions."""
 import dataclasses
 from abc import abstractmethod, ABC
-from functools import partial
 from typing import Dict, Iterable, List, Optional, Sequence, Tuple, TYPE_CHECKING
 
 import numpy as np
@@ -648,24 +647,15 @@ def _fit_exponential_decay(
     layer_fid_0 = np.clip(np.exp(slope), 0, 1)
     a_0 = np.clip(np.exp(intercept), 0, 1)
 
-    def do_curve_fit():
-        curve_fit = partial(
-            optimize.curve_fit,
+    try:
+        (a, layer_fid), pcov = optimize.curve_fit(
             exponential_decay,
             cycle_depths,
             fidelities,
             p0=(a_0, layer_fid_0),
             bounds=((0, 0), (1, 1)),
+            maxfev=1000,
         )
-        try:
-            return curve_fit()
-        except RuntimeError:  # pragma: no cover
-            # Curve_fit didn't find a solution. Try once more w/ higher maxfev.
-            # Default (in SciPy v.1.14) is 100*(1+len(p0)) = 300 for our p0.
-            return curve_fit(maxfev=1000)
-
-    try:
-        (a, layer_fid), pcov = do_curve_fit()
     except ValueError:  # pragma: no cover
         return 0, 0, np.inf, np.inf
 

